input-profile.ts

```diff
@@ -0,0 +1,360 @@
+
+    /**
+     * An optional folder path for loading custom XR input profiles.
+     */
+    @property.string()
+    customProfileFolder!: string;
+
+    /**
+     * The default 3D controller model used when a custom model fails to load.
+     */
+    @property.object()
+    defaultController!: Object3D;
+
+    /**
+     * The object which has HandTracking component added to it.
+     */
+    @property.object()
+    trackedHand!: Object3D;
+
+    /**
+     * If true, the input profile will be mapped to the default controller, and no dynamic 3D model of controller will be loaded.
+     */
+    @property.bool(false)
+    mapToDefaultController!: boolean;
+
+    /**
+     * If true, adds a VR mode switch component to the loaded controller model.
+     */
+    @property.bool(true)
+    addVrModeSwitch!: boolean;
+
+    onActivate() {
+        this._handedness = hands[this.handednessIndex];
+        const defaultHandName =
+            'Hand' + this._handedness.charAt(0).toUpperCase() + this._handedness.slice(1);
+        this.trackedHand =
+            this.trackedHand || this.object.parent.findByNameRecursive(defaultHandName)[0];
+        console.log(this.trackedHand);
+        this.defaultController = this.defaultController || this.object.children[0];
+        this._defaultControllerComponents = this._getComponents(this.defaultController);
+
+        this.engine.onXRSessionStart.add(() => {
+            this.engine.xr?.session.addEventListener(
+                'inputsourceschange',
+                this._onInputSourcesChange.bind(this)
+            );
+        });
+    }
+
+    onDeactivate() {
+        this.engine.xr?.session?.removeEventListener(
+            'inputsourceschange',
+            this._onInputSourcesChange.bind(this)
+        );
+    }
+
+    /**
+     * Sets newly loaded controllers for the HandTracking component to proper switching.
+     * @param controllerObject The controller object.
+     * @hidden
+     */
+
+    private _setHandTrackingControllers(controllerObject: Object3D) {
+        const handtrackingComponent: HandTracking | null =
+            this.trackedHand.getComponent(HandTracking);
+        if (!handtrackingComponent) return;
+        handtrackingComponent.controllerToDeactivate = controllerObject;
+    }
+
+    /**
+     * Retrieves all components from the specified object and its children.
+     * @param obj The object to retrieve components from.
+     * @return An array of components.
+     * @hidden
+     */
+    private _getComponents(obj: Object3D | null) {
+        const components: Component[] = [];
+        if (obj == null) return components;
+        const stack = [obj];
+        while (stack.length > 0) {
+            const currentObj = stack.pop();
+            const comps = currentObj!
+                .getComponents()
+                .filter((c: Component) => !this.toFilter.has(c.type));
+            components.push(...comps);
+            const children = currentObj!.children || [];
```

Comment:
`|| []` can't happen, this getter **always** returns something, so you can just do:

```ts
const children = currentObj.children;
```